### DISCUSSION TITLE
Async db manager initializer and GRDBQuery

### DISCUSSION THREAD

**Discussion by Jason-Abbott** [groue/GRDB.swift#1668] [Q&A] [answered]
My needs are met by the [@Query Demo App](https://github.com/groue/GRDBQuery/tree/main/Documentation/QueryDemo) except that my [PlayerRepository](https://github.com/groue/GRDBQuery/blob/main/Documentation/Players/Sources/Players/PlayerRepository.swift) equivalent has an asynchronous initializer which isn’t usable, as far as I can tell, as a [DatabaseContext](https://swiftpackageindex.com/groue/grdbquery/0.10.1/documentation/grdbquery/databasecontext).

It is asynchronous because it applies migrations from a `@globalActor` “Registry”.

I am a Swift app novice so there’s a reasonable chance I settled on anti-pattern 🙂 but what I have are basically these Xcode targets:

- App
- SharedLib (has the *Registry*)
- FeatureLib1
- FeatureLib2
- FeatureLib3

`App` depends on everything else, `SharedLib` depends on nothing (except GRDB), and the `FeatureLib`s depend only on `SharedLib`.

Each `FeatureLib` registers its own unique migrations (and various other non-GRDB things). This all works fine in unit tests but I’m wondering if I’ll need to approach it differently in order to use `DatabaseContext` and `GRDBQuery`.

(relevant code:)
```swift
private init(_ connection: any DatabaseWriter) async throws {
   self.connection = connection
   try await migrate()
}

private func migrate() async throws {
   var migrator = DatabaseMigrator()
   for m in coreMigrations { m.register(with: &migrator) }
   for m in await ProviderRegistry.migrations { m.register(with: &migrator) }
   try migrator.migrate(connection)
}
```

**Comment by Jason-Abbott** [comment]
Maybe the easy answer is to dispense with the *registry* notion and allow `App`, since it imports everything, to iterate the `FeatureLib`s in some sense to combine their migrations for the `Repository` without an `Actor` in the chain.

Barring a quick alternate solution for `DatabaseContext`, I’ll probably start down this path ...

**Comment by groue** [comment, marked as answer]
Hello @Jason-Abbott,

Is there something that prevents your app from waiting for the database to be migrated before registering the `DatabaseContext`?

Here is a quick and dirty sample code:

```swift
import GRDB
import GRDBQuery

struct LoaderView: View {
    // Initially nil, then not nil when the database is ready
    @State var databaseContext: DatabaseContext?
    
    var body: some View {
        Group {
            if let databaseContext {
                MyView().databaseContext(databaseContext)
            } else {
                ProgressView()
            }
        }
        .task {
            do {
                let dbQueue = try DatabaseQueue()
                try await Task.sleep(for: .seconds(1)) // migrate
                databaseContext = DatabaseContext.readWrite { dbQueue }
            } catch {
                databaseContext = DatabaseContext.readWrite { throw error }
            }
        }
    }
}

struct MyView: View {
    @Environment(\.databaseContext) var databaseContext
    
    var body: some View {
        let value = try! databaseContext.reader.read { db in "Connected!" }
        Text(value)
    }
}

#Preview {
    LoaderView()
}
```

**Comment by Jason-Abbott** [comment]
@groue I think that should work perfectly. Thank you!

**Reply by groue** [reply]
Great! Please come back if something does not work as expected :-)

**Comment by Jason-Abbott** [comment]
I am having trouble convincing an `NSViewRepresentable` to use the `databaseContext` I’m assigning to the `#preview` of its parent `View`. The `NSViewRepresentable` generates content for an `NSTextView` using its own 
```swift
@Environment(\.databaseContext) var databaseContext
```
Print statements confirm the `#preview` `View` has created an in-memory database with test content and assigned it to `@Environment(\.databaseContext)`.

Behind-the-scenes, the `#preview` still loads *its* parent `@main App` with the `.task` that creates a `databaseContext`, as you suggested above.  And *that* is the in-memory database instance (empty) the `NSViewRepresentable` sees.

I am trying to keep this brief, and not paste in much code, since I don’t think it’s a GRDB issue. I’m just curious if there’s anything special I should consider when mixing `#preview`s with an `@Environment` database context.

**Reply by Jason-Abbott** [reply]
@groue maybe hold on putting any time into this unless you have a quick thought. I’m streamlining my use of `@Environment` to reduce interdependencies. It’s too hard to reason about what I have going on. I’m sure that will fix everything. 🤨

**Reply by groue** [reply]
Hi @Jason-Abbott. I think you observe an expected behavior: the app view overrides the context present in the environment, which is discarded.

You can instruct the view that overrides the context to avoid doing so. If I take the `LoaderView` above:

```diff
+#if DEBUG
+extension EnvironmentValues {
+    // A database context that should be preferred, when not nil.
+    @Entry var previewDatabaseContext: DatabaseContext?
+}
+#endif
 
 struct LoaderView: View {
+    #if DEBUG
+    // Allow previews to provide a custom database context
+    @Environment(\.previewDatabaseContext) var previewDatabaseContext
+    #endif
 
     // Initially nil, then not nil when the database is ready
     @State var databaseContext: DatabaseContext?
 
     var body: some View {
         Group {
             if let databaseContext {
                 MyView().databaseContext(databaseContext)
             } else {
                 ProgressView()
             }
         }
         .task {
+            #if DEBUG
+            if let previewDatabaseContext {
+                databaseContext = previewDatabaseContext
+                return
+            }
+            #endif
 
             do {
                 let dbQueue = try DatabaseQueue()
                 try await Task.sleep(for: .seconds(1)) // migrate
                 databaseContext = DatabaseContext.readWrite { dbQueue }
             } catch {
                 databaseContext = DatabaseContext.readWrite { throw error }
             }
         }
     }
 }
```

Usage:

```swift
#Preview {
    LoaderView()
        .environment(\.previewDatabaseContext, .readWrite { ... })
}
```

**Reply by Jason-Abbott** [reply]
You are right. I actually realized as much as I was on a nice bicycle ride the same day but didn’t get back here before a holiday with my daughter.

I ended up refactoring my use of `.environment` along with `#if DEBUG` seasoning to eliminate duplicate instances while previewing.
